### ISSUE TITLE
The `SequenceStateBuilder` needs to be improved

### ISSUE THREAD

**Issue by probberechts** [PySport/kloppy#385]
I think the logic for building sequences could be improved. Below are a few issues that I noticed:
- Typically, an event stream starts with a few off-ball events such as an event that announces the lineups, the start of a period, etc. These events should not be part of the possession sequence of a team.
- Events in which the goal keeper picks up or claims the ball should be able to start a new sequence
- Take ons should be able to start a new sequence
- Successful interceptions should be able to start a new sequence
- A single failed pass should not start a new sequence
- When a sequence ends with a duel between two players, the lost duel should be part of the old sequence and the winning duel should be part of the new sequence
- A shot should not always end a sequence. For example, a rebound should be part of the same sequence.

**Comment by probberechts** [comment]
While looking up some definitions, I figured Opta uses both the concept of a "sequence" and a "possession".

> events are organised into sequences and possessions.
> **Sequences** are defined as passages of play which belong to one team and are ended by defensive actions, stoppages in play or a shot.
> **Possessions** are defined as one or more sequences in a row belonging to the same team. A series of passes leading to a shot which is saved and results in a corner kick would comprise one possession since the same team retains control, but more than one sequence, since the ball has gone out of play. A possession is ended by the opposition gaining control of the ball.
> -- https://www.statsperform.com/resource/introducing-a-possessions-framework

Hence, maybe we should keep the `SequenceStateBuilder` as is and add a `PossessionStateBuilder`?

**Comment by DriesDeprest** [comment]
Yes, I think it would be wise to make the distinction and use the definitions used by Opta.

**Comment by koenvo** [comment]
Would be great. Not sure how difficult this will make things, but maybe good to be able to pass some configuration

**Comment by lodevt** [comment]
Hi!

I agree that adding a `PossessionStateBuilder` alongside the `SequenceStateBuilder` is a great idea. However, based on the issues you noticed, I still believe a few adjustments to the `SequenceStateBuilder` could be beneficial.

---

> * Typically, an event stream starts with a few off-ball events such as an event that announces the lineups, the start of a period, etc. These events should not be part of the possession sequence of a team.  
> 
**→** I think we should indeed exclude these events (both in sequences and possessions).

> * Events in which the goalkeeper picks up or claims the ball should be able to start a new sequence.  
> * Take-ons should be able to start a new sequence.  
> * Successful interceptions should be able to start a new sequence.  
> 
**→** Even if we introduce a new `PossessionStateBuilder`, I think these events should still be able to start a new sequence.

> * A single failed pass should not start a new sequence.  
> 
**→** I would suggest keeping the sequence logic as is in this case. So these would be two **sequences** belonging to the same **possession**.

> * When a sequence ends with a duel between two players, the lost duel should be part of the old sequence and the winning duel should be part of the new sequence.  
>
 **→** I think we should follow this rule for both sequences and possessions.

> * A shot should not always end a sequence. For example, a rebound should be part of the same sequence.  
> 
**→** This one falls a bit in a grey zone for me. Maybe we can keep it as two sequences (again belonging to the same possession)?

---

Let me know what you think!

**Comment by probberechts** [comment]
> Hi!
> 
> I agree that adding a `PossessionStateBuilder` alongside the `SequenceStateBuilder` is a great idea. However, based on the issues you noticed, I still believe a few adjustments to the `SequenceStateBuilder` could be beneficial.
> 
> > * Typically, an event stream starts with a few off-ball events such as an event that announces the lineups, the start of a period, etc. These events should not be part of the possession sequence of a team.

I agree. 

> 
> **→** I think we should indeed exclude these events (both in sequences and possessions).
> 
> > * Events in which the goalkeeper picks up or claims the ball should be able to start a new sequence.
> > * Take-ons should be able to start a new sequence.
> > * Successful interceptions should be able to start a new sequence.
> 
> **→** Even if we introduce a new `PossessionStateBuilder`, I think these events should still be able to start a new sequence.

Indeed. Moreover, defensive actions that are followed by a controlled action like a pass or dribble should be counted as the start of a sequence.

> > * A single failed pass should not start a new sequence.
> 
> **→** I would suggest keeping the sequence logic as is in this case. So these would be two **sequences** belonging to the same **possession**.

Yes.

> > * When a sequence ends with a duel between two players, the lost duel should be part of the old sequence and the winning duel should be part of the new sequence.
> 
> **→** I think we should follow this rule for both sequences and possessions.

Indeed.

> > * A shot should not always end a sequence. For example, a rebound should be part of the same sequence.
> 
> **→** This one falls a bit in a grey zone for me. Maybe we can keep it as two sequences (again belonging to the same possession)?

I think keeping it as two sequences would make the most sense.

> Let me know what you think!

In general, I would try to reproduce Opta's approach here. They provide a sequenceId and possessionId in the MA13 feed. Let me know if you would like to work on this. I can share some sample data.

**Comment by probberechts** [comment]
Great! This is the [MA13 feed](https://github.com/user-attachments/files/20862479/statsperform_event_ma13.json) corresponding to [our current test files](https://github.com/PySport/kloppy/blob/master/kloppy/tests/files/statsperform_event_ma1.json). Look for the `"sequenceId"` and `"possessionId"` keys.
